src/Microsoft.Diagnostics.Runtime/src/Common/IFileLocator.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Diagnostics.Runtime
+{
+    /// <summary>
+    /// An implementation of 
+    /// </summary>
+    public interface IFileLocator
+    {
+        /// <summary>
+        /// Locates a PE Image archived under the given properties.
+        /// </summary>
+        /// <param name="fileName">The file name or path of the binary to locate.</param>
+        /// <param name="buildTimeStamp">The build timestamp the binary is indexed under.</param>
+        /// <param name="imageSize">The image size the binary is indexed under.</param>
+        /// <param name="checkProperties">Whether or not to validate the properties of the binary after download.</param>
+        /// <returns>A full path on disk (local) of where the binary was copied to or <see langword="null"/> if it was not found.</returns>
+        string? FindPEImage(string fileName, int buildTimeStamp, int imageSize, bool checkProperties);
+
+        /// <summary>
+        /// Locates a PE Image potentially archived under an ELF or Mach-O binary's properties.
+        /// </summary>
+        /// <param name="fileName">The file name or path of the binary to locate.</param>
+        /// <param name="archivedUnder">The file or keyword that this binary is archived under.</param>
+        /// <param name="buildIdOrUUID">The buildId or UUID of the binary specified by <paramref name="archivedUnder"/>.</param>
+        /// <param name="originalPlatform">The platform of the binary specified by <paramref name="archivedUnder"/>.</param>
+        /// <param name="checkProperties">Whether or not to validate the properties of the binary after download.</param>
+        /// <returns>A full path on disk (local) of where the binary was copied to or <see langword="null"/> if it was not found.</returns>
+        string? FindPEImage(string fileName, string archivedUnder, ImmutableArray<byte> buildIdOrUUID, OSPlatform originalPlatform, bool checkProperties);
+
+        /// <summary>
+        /// Locates an Elf binary.
+        /// </summary>
+        /// <param name="fileName">The file name or path of the binary to locate.</param>
+        /// <param name="archivedUnder">The file or keyword that this binary is archived under, <see langword="null"/> if its archived under its own properties.</param>
+        /// <param name="buildId">The buildId of the Elf image to locate or the buildId of the image specified by <paramref name="archivedUnder"/>.</param>
+        /// <param name="checkProperties">Whether or not to validate that the given file matches the build id.</param>
+        /// <returns>A full path on disk (local) of where the binary was copied to or <see langword="null"/> if it was not found.</returns>
+        string? FindElfImage(string fileName, string? archivedUnder, ImmutableArray<byte> buildId, bool checkProperties);
```

Comment:
I'm not sure about exposing "archivedUnder" part of the index in this public API.  The "coreclr" is part of symbol server protocol.  I think of this more abstractly as either it is an "Identity" (the module itself), "Runtime" (the coreclr special keys) or "Symbol" index/keys.  A enum might be a better approach.

I guess related to this is that this interface couldn't be used to download symbols (PDB or dwarf for Linux/MacOS). I'm sure this is intentional and a ISymbolFileLocator could be added if clrmd ever needs to download them.

Reply:
> I'm not sure about exposing "archivedUnder" part of the index in this public API. The "coreclr" is part of symbol server protocol.

I agree it's odd, it just felt like the least-odd option.  I know there's not currently anything other than `coreclr` as the value, but it still feels weird not having `archivedUnder` and just silently assuming it's `coreclr`.

If we don't include it as a parameter here then it becomes a "hidden" part of the API.  For example, we'd need something like `FindElfImageUnderClrProperties(string fileName, string? archivedUnder, ImmutableArray<byte> buildId, bool checkProperties);` to make it clear that we would be using "coreclr" there.  (Because FindElfImage with that overload already exists.)  Which also feels weird.

> I guess related to this is that this interface couldn't be used to download symbols (PDB or dwarf for Linux/MacOS). I'm sure this is intentional and a ISymbolFileLocator could be added if clrmd ever needs to download them.

Correct, we don't support downloading anything other than images, for now.

Reply:
I was suggesting replacing the string archiveUnder with a more abstract enum not removing it or assuming it is coreclr. But I guess the only caller of this interface is clrmd itself.  There might be other implementations outside clrmd, but they can can do the same thing as clrmd's impl.

Reply:
Ahh interesting, I'll take a look at doing that.